### PR TITLE
Fix invalid while_let_on_iterator suggestion

### LINKED ISSUE #3780: Bad `while_let_on_iterator` suggestion when pattern is refutable
Here is an example:

```rust
fn main() {
    let v = vec![1, 2, 3];
    
    {
        let mut it = v.windows(2);
        while let Some([x, y]) = it.next() {
            println!("x: {}", x);
            println!("y: {}", y);
        }
    }
    /*
    warning: this loop could be written as a `for` loop
     --> src/main.rs:6:34
      |
    6 |         while let Some([x, y]) = it.next() {
      |                                  ^^^^^^^^^ help: try: `for [x, y] in it { .. }`
      |
      = note: #[warn(clippy::while_let_on_iterator)] on by default
    */
    
    // The Clippy suggestion fails to compile
    /*
    {
        let mut it = v.windows(2);
        for [x, y] in it {
            println!("x: {}", x);
            println!("y: {}", y);
        }
    }
    */
}
```
[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=731f85b2f9fdc0bb2e4147523048b101)

### PR DESCRIPTION
**Pull request by jfirebaugh**
Slice patterns without `..` are refutable. (Well, except if the match source is an array type -- does `is_refutable` have enough information to check this though?)

Fixes #3780.

### CODE REVIEW

File: clippy_lints/src/utils/mod.rs [hunk 1]
```diff
@@ -958,7 +958,11 @@ pub fn is_refutable(cx: &LateContext<'_, '_>, pat: &Pat) -> bool {
             }
         },
         PatKind::Slice(ref head, ref middle, ref tail) => {
-            are_refutable(cx, head.iter().chain(middle).chain(tail.iter()).map(|pat| &**pat))
+            if middle.is_none() {
+                true
```
Comment (oli-obk) [inline clippy_lints/src/utils/mod.rs]: It might still be irrefutable if `head` or `tail` are irrefutable

**Comment by oli-obk** [comment]
> does `is_refutable` have enough information to check this though?)

Not sure, but we can probably pass in the relevant information

**Comment by mikerite** [comment]
The patterns `[a,..]` and `[..,b]` require at least 1 element so they won't match `[]`. This makes me think that the only irrefutable slice pattern is `[..]` so maybe just check for that.

File: clippy_lints/src/utils/mod.rs (hunk 1 above)
Comment (jfirebaugh) [reply clippy_lints/src/utils/mod.rs]: Are you referring to matches on array types? As @mikerite points out, the only irrefutable slice pattern on slices is `[..]`.

**Comment by jfirebaugh** [comment]
> does `is_refutable` have enough information to check this though?

It appears so; I've added type-checking in the latest commit, though I am not too sure of its correctness. I see that MIR [handles the case where the pattern is `ty::Ref`](https://github.com/rust-lang/rust/blob/0dbb867beb187119bd0bca1f05de6b00cd72f141/src/librustc_mir/hair/pattern/mod.rs#L518-L526), as well as `ty::Error`. Does anyone know in what situations these arise?

As an aside, it would be nice to lean on the compiler for something like `is_refutable`. What is the stance toward making items public in, say, `rustc_mir` solely for the benefit of external tooling like clippy? For example, something like [this](https://github.com/rust-lang/rust/commit/4d8cc91ba7ba0a63a0738decdc7da0eb4923ad2a) would presumably be more robust than clippy's current implementation.

**Comment by flip1995** [comment]
Ping @jfirebaugh. I'm going over old PRs, that were abandoned by us reviewers (sorry for that!) or by the authors. Are you still interested in completing this?

**Comment by flip1995** [comment]
Ping from triage @jfirebaugh. Sorry for neglecting this PR and thanks for your contribution. If you want to continue working on this, feel free to reopen and ping me. I'll be happy to help you wrapping this up.

**Comment by HKalbasi** [comment]
finished in #5559
@rustbot label -S-inactive-closed

**Pull request closed without merging**